river/KeyboardHandler.zig

```diff
@@ -0,0 +1,113 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2020 - 2022 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const std = @import("std");
+const assert = std.debug.assert;
+const wlr = @import("wlroots");
+const wl = @import("wayland").server.wl;
+const xkb = @import("xkbcommon");
+
+const log = std.log.scoped(.keyboard_handler);
+
+const KeycodeSet = @import("KeycodeSet.zig");
+const Seat = @import("Seat.zig");
+
+const server = &@import("main.zig").server;
+
+/// Pressed keys for which a mapping was triggered on press
+eaten_keycodes: KeycodeSet = .{},
+
+pub fn handleKey(self: *Self, seat: *Seat, wlr_device: *wlr.InputDevice, event: *wlr.Keyboard.event.Key) void {
```

Comment:
I don't think we actually need to factor this logic out of `Keyboard.zig`. Instead we could pass the `wlr_keyboard_group`'s "virtual" input device to `Seat.addDevice()`, creating a river `Keyboard` for the `wlr_keyboard_group`'s aggregate keyboard and using the exact same code paths.

What do you think of that approach?

Reply:
I was initially convinced this would not work, hence the  current implementation. But can not remember why, so I'll give it a try.